quil-plotting/quil_plotting/dataframe.py

```diff
@@ -0,0 +1,346 @@
+from quil.expression import Expression, PrefixExpression, PrefixOperator
+from quil.instructions import Gate, Pragma, Pulse, Qubit, ShiftPhase
+from quil.program import Program
+
+
+# color
+class Color(str, Enum):
+    GREEN = "#00b5ad"  # Teal
+    YELLOW = "#ffc504"  # Yellow
+    RED = "#ef476f"  # Magneta
+    BLUE = "#3d47d9"  # Palatinate Blue
+    GRAY = "#8a8b92"  # Gray
+    NAVY = "#0d0d36"  # Cetacean Blue
+
+
+def hex_to_rgba(hex_color: str, opacity: float = 1.0) -> str:
+    """Convert a hex color to an rgba string.
+
+    >>> hex_to_rgba("#00B5AD")
+    'rgba(0,181,173,1.0)'
+
+    :param hex_color: The hex color code.
+    :param opacity: The opacity for the rgba color.
+    """
+    hex_color = hex_color.strip("#")
+    r, g, b = tuple(int(hex_color[i : i + 2], 16) for i in (0, 2, 4))
+    return f"rgba({r},{g},{b},{opacity})"
+
+
+def rgba_to_hex(rgba: Union[np.ndarray, List[int]]) -> str:
+    """Convert an RGBA array to a hex color string.
+
+    :param rgba: Array or list containing RGB components on [0-255]. The alpha component is optional.
+    :return: Hex color string.
+    """
+    if len(rgba) == 3:
+        return f"#{rgba[0]:02x}{rgba[1]:02x}{rgba[2]:02x}"
+    elif len(rgba) == 4:
+        return f"#{rgba[0]:02x}{rgba[1]:02x}{rgba[2]:02x}{rgba[3]:02x}"
+    else:
+        raise ValueError("RGBA array must have 3 or 4 elements")
+
+
+def expand_with_source_mapping(program: Program, include_pragmas: bool = False) -> Tuple[Program, List[int]]:
+    """Expand any instructions in the program which have a matching calibration, leaving the others
+    unchanged.
+
+    :param program: A quil.Program.
+    :return: A quil.Program with the instructions expanded and a source map.
+    The source map is a list with the length of the expanded instructions, and indicates the index
+    of the logical instruction in the original program from which the expanded instruction originated.
+    """
+    instructions = program.body_instructions
+    calibrations = program.calibrations
+    expanded_program = program.clone_without_body_instructions()
+
+    source_map = []
+    expanded_instructions = []
+    for logical_index, instruction in enumerate(instructions):
+        expanded_instruction = calibrations.expand(instruction, [])
+        if expanded_instruction is None:
+            expanded_instruction = [instruction]
+        if include_pragmas is False:
+            expanded_instruction = [inst for inst in expanded_instruction if not isinstance(inst.inner(), Pragma)]
+        source_map += [logical_index] * len(expanded_instruction)
+        expanded_instructions += expanded_instruction
+    expanded_program.add_instructions(expanded_instructions)
+    return expanded_program, source_map
+
+
+def add_plot_metadata(
+    df: pd.DataFrame,
+    runners: str = "Qubit",
+    color_by: str = "Channel Type",
+    runner_order: str = "Time (s)",
+    exclude_readout: bool = True,
+    normalize_by: Optional[str] = "Channel Type",
```

Comment:
Instead of plain strings, can we make "categoricals" and the like enums?

Reply:
Yes, however I'm not sure if using enums with a pandas dataframe is entirely conventional. I'm not actually sure what best practice would be here.